lang/py/avro/io.py

```diff
@@ -199,13 +201,745 @@ def _map_iterator(node: ValidationNode) -> ValidationNodeGeneratorType:
+            self, datum_reader: "DatumReader", writers_schema: avro.schema.RecordSchema, readers_schema: avro.schema.RecordSchema) -> Mapping[str, object]:
+        pass
+
+
+class DatumReader:
+    """Deserialize Avro-encoded data into a Python data structure."""
+
+    _writers_schema: Optional[avro.schema.Schema]
+    _readers_schema: Optional[avro.schema.Schema]
+
+    def __init__(self, writers_schema: Optional[avro.schema.Schema] = None, readers_schema: Optional[avro.schema.Schema] = None) -> None:
+        """
+        As defined in the Avro specification, we call the schema encoded
+        in the data the "writer's schema", and the schema expected by the
+        reader the "reader's schema".
+        """
+        self._writers_schema = writers_schema
+        self._readers_schema = readers_schema
+
+    @property
+    def writers_schema(self) -> Optional[avro.schema.Schema]:
+        return self._writers_schema
+
+    @writers_schema.setter
+    def writers_schema(self, writers_schema: avro.schema.Schema) -> None:
+        self._writers_schema = writers_schema
+
+    @property
+    def readers_schema(self) -> Optional[avro.schema.Schema]:
+        return self._readers_schema
+
+    @readers_schema.setter
+    def readers_schema(self, readers_schema: avro.schema.Schema) -> None:
+        self._readers_schema = readers_schema
+
+    def read(self, decoder: Decoder) -> object:
+        if self.writers_schema is None:
+            raise avro.errors.IONotReadyException("Cannot read without a writer's schema.")
+        if self.readers_schema is None:
+            self.readers_schema = self.writers_schema
+        return self.read_data(self.writers_schema, self.readers_schema, decoder)
+
+    def read_data(self, writers_schema: avro.schema.Schema, readers_schema: avro.schema.Schema, decoder: Decoder) -> object:
+        # schema matching
+        if not readers_schema.match(writers_schema):
+            raise avro.errors.SchemaResolutionException("Schemas do not match.", writers_schema, readers_schema)
+
+        logical_type = getattr(writers_schema, "logical_type", None)
+
+        # function dispatch for reading data based on type of writer's schema
+        if isinstance(writers_schema, avro.schema.UnionSchema) and isinstance(readers_schema, avro.schema.UnionSchema):
+            return self.read_union(writers_schema, readers_schema, decoder)
+
+        if isinstance(readers_schema, avro.schema.UnionSchema):
+            # schema resolution: reader's schema is a union, writer's schema is not
+            for s in readers_schema.schemas:
+                if s.match(writers_schema):
+                    return self.read_data(writers_schema, s, decoder)
+
+            # This shouldn't happen because of the match check at the start of this method.
+            raise avro.errors.SchemaResolutionException("Schemas do not match.", writers_schema, readers_schema)
+
+        if writers_schema.type == "null":
+            return None
```

Comment:
Even while returning `None` an encoder can potentially read something from the stream
```suggestion
            return decoder.read_null()
```